pkg/sql/sqlliveness/slinstance/slinstance.go

```diff
@@ -237,6 +239,58 @@ func (l *Instance) extendSession(ctx context.Context, s *session) (bool, error)
 	return true, nil
 }
 
+// TimeoutError is used to indicate when a call to createSessionWithTimeout or
+// extendSessionWithTimeout has failed due to timeout.
+type TimeoutError struct{}
+
+func (t TimeoutError) Error() string {
+	return "session timeout"
+}
+
+var _ error = &TimeoutError{}
+
+func (l *Instance) createSessionWithTimeout(
+	ctx context.Context, timeout time.Duration,
+) (*session, error) {
+	var s *session
+	var err error
+	var createChan = make(chan struct{})
+	go func() {
+		s, err = l.createSession(ctx)
+		close(createChan)
+	}()
+
+	t := timeutil.NewTimer()
+	t.Reset(timeout)
+	select {
+	case <-t.C:
+		return nil, &TimeoutError{}
+	case <-createChan:
+		return s, err
+	}
+}
```

Comment:
1) this seems complex
2) This fails to actually cancel the underlying work, which, I think, is bad

Can we instead with `context.WithTimeout` and then detect the deadline exceeded error?

Reply:
Changed to use `context.WithTimeout`